data/CourseMemoryDAO.test.js

```diff
@@ -6,14 +7,237 @@ const chai = require('chai');
 
 const should = chai.should();
 const CourseMemoryDAO = require('./CourseMemoryDAO');
+const InMemoryDataStore = require('./InMemoryDataStore');
+
+const Course = require('./Course');
 
 describe('CourseMemoryDAO', () => {
   describe('construction', () => {
-    it('Should initialize with an empty course array and a maxCourseId of 0', () => {
-      /*const dao = new CourseMemoryDAO();
-      chai.expect(dao.courses).to.be.instanceOf(Array);
-      chai.except(dao.courses).to.be.empty();
-      chai.expect(dao.maxCourseId).to.be.equal(0);*/
+    it('Should initialize with access to the in-memory data store object', () => {
+      const dao = new CourseMemoryDAO();
+      chai.expect(dao.data).to.be.equal(InMemoryDataStore);
+    });
+
+    it('Should not be affected by parameters', () => {
+      const dao = new CourseMemoryDAO([1, 2, 3], false, 'test');
+      chai.expect(dao.data).to.be.equal(InMemoryDataStore);
+    });
+  });
+  describe('newCourse()', () => {
+    let dao;
+    beforeEach(() => {
+      dao = new CourseMemoryDAO();
+      dao.data.clear();
+    });
+
+    afterEach(() => {
```

Comment:
I'm not sure you need this because you set it to a new instance every time in the before each section. If you didn't reassign it then this would be necessary.

Reply:
The reason I think I need to clear it is because the data is shared between all DAO's. So if a test add's 3 courses to the in memory data, then the next test add's 1 course and checks to see if the courses array has a length of 1 it will fail (because there are now 4 courses in the courses array). I guess the alternative would be to write the tests so they don't rely on the in-memory data starting with an empty state.

Reply:
yes but you clear it in the beforeEach as well.

```
1. Test 1 runs before each (cleared dao is cleared)
-- dao is cleared for test--
2. Test 1 runs
3. Test 1 runs afterEach (dao is cleared)
- dao is now cleared -
4. Test 2 runs beforeEach (cleared dao is cleared again)
5. Test 2 runs 
6. Test 2 runs afterEach (dao is cleared)
--- dao is now cleared ---

You can remove the clear from the beforeEach, because each test clears it after it completes.